### PR TITLE
Use golang rpm to build singularity rpm

### LINKED ISSUE #2350: Singularity RPM spec file hard codes amd64 arch for go
## Version of Singularity:
3.0.1

### Actual behavior
The singularity.spec file hard codes `amd64` for the architecture when downloading Go. We should have a way that the architecture can be changed when doing the build of the RPM. For example:

```
rpmbuild --define '_goarch ppc64le' [...]
```

`%{_arch}` cannot be used as that returns x86_64 and not amd64.

### PR DESCRIPTION
**Pull request by DrDaveD**
**Description of the Pull Request (PR):**

Converts rpm building to use an existing golang package instead of downloading an architecture-specific go language version.  This is only recently possible because RHEL 7.6 removed golang by default, so epel could add it back.  It is currently at version 1.11.2.

**This fixes or addresses the following GitHub issues:**

- Fixes #2350

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Review by jmstover — APPROVED** [review]
LGTM

**Comment by DrDaveD** [comment]
Why is the "test-compile" check still waiting.  Is that CircleCI?

**Comment by DrDaveD** [comment]
Closing & reopening to try to kick the circle CI test (after removing the CircleCI webhook on my own fork).

**Comment by DrDaveD** [comment]
Trying close & open again after unfollowing my fork in CircleCI.

**Comment by jscook2345** [comment]
@DrDaveD @jmstover 

It looks like this causes RPM builds for opensuse to fail (golang is needed by...). opensuse leap does not have golang1.11 available to zypper at the moment. I'm looking to see if there is a way to include it.

**Comment by jscook2345** [comment]
So I found a way to include go1.11 with opensuse, but it is OS specific.

<pre>
sudo zypper ar http://download.opensuse.org/repositories/devel:/languages:/go/openSUSE_Leap_15.0/devel:languages:go.repo
sudo zypper refresh
zypper info go1.11
sudo zypper install go1.11
</pre>

**Comment by jscook2345** [comment]
@DrDaveD  @jmstover 

If you could put in some suse specific checks (or maybe we should have two spec files?) I think we can merge and then add the suse-specific info afterwards.

Everyone agree?

**Comment by DrDaveD** [comment]
I agree.  I would vote for putting in suse-specific code in the same spec.

**Comment by jmstover** [comment]
@DrDaveD This is totally not the best, but what about something like:
```
%if "%{_target_vendor}" == "suse"
    sudo zypper ar http://download.opensuse.org/repositories/devel:/languages:/go/openSUSE_Leap_15.0/devel:languages:go.repo
    sudo zypper refresh
    zypper info go1.11
    sudo zypper install go1.11
%endif
```

I really don't think that will work for SLES 11 / 12 though ...

**Comment by DrDaveD** [comment]
If that works it's OK with me.  Presumably you'd have to exclude the Requires: golang line as well.

**Comment by jscook2345** [comment]
> I really don't think that will work for SLES 11 / 12 though ... 

It may not, but it does have an devel repo from 11.4 through 15.0 for SLES. 

    http://download.opensuse.org/repositories/devel:/languages:/go/

**Comment by DrDaveD** [comment]
Is zypper just the yum of suse?  If so, can you just use a different Requires statement?

**Comment by jscook2345** [comment]
@DrDaveD 

Yes, `zypper` similar to `yum` for suse.

**Comment by DrDaveD** [comment]
@jscook2345 @jmstover Can we go ahead and merge this one?

**Comment by jscook2345** [comment]
> @jscook2345 @jmstover Can we go ahead and merge this one?

@DrDaveD I think we need to add the checks for opensuse. Though I am wondering if perhaps we should have two spec files. One for CentOS/RHEL/Fedora and one for OpenSUSE/SLES.

@jmstover thoughts?

**Comment by DrDaveD** [comment]
Certainly there should be no need for a separate .spec file.  It sounds like all that's needed is a simple if statement with a different Requires.  I thought we had agreed to put in the suse-specific pieces after this was merged, but I'll see if I can get it to build under docker.

**Comment by DrDaveD** [comment]
Ok that was easy.  I added a %if around the BuildRequires: golang to just have BuildRequires: go on suse, and got it to build, install, and run under the opensuse:latest docker image.

**Comment by DrDaveD** [comment]
It installed go1.10, which doesn't pass a go fmt check, but that isn't required to build and run the rpm.

File: dist/rpm/singularity.spec.in
Comment (jmstover) [inline dist/rpm/singularity.spec.in:42]: This will never be satisfied ... I can't find a go package on the SLES 11 or 12 VM's I have available to me.

Comment (DrDaveD) [reply dist/rpm/singularity.spec.in:42]: Ok, there appears to be no sles docker hub images (deliberately according to [this](https://www.suse.com/documentation/sles-12/singlehtml/book_sles_docker/book_sles_docker.html#docker.building.images)) so I cannot verify it, but according to this [documentation](https://en.opensuse.org/SDB:Go#Getting_Started_.2F_Installation) you need to add the repository at [devel:languages:go](https://build.opensuse.org/project/show?project=devel:languages:go).  I don't know if there's a better way to do it, but if you click on the "Repositories" tab and click through to the relevant download repository for your sles version you get to a ".repo" file that you can import with `zypper ar -r`, for example 
```
zypper ar -r https://download.opensuse.org/repositories/devel:/languages:/go/SLE_12_SP4/devel:languages:go.repo
```
This is pretty similar to what @jscook2345 said above but it allows you to directly install go with `zypper install go`.

**Comment by jscook2345** [comment]
@DrDaveD I think we're good to go if you want to resolve the conflicts first.

**Comment by DrDaveD** [comment]
@jscook2345 Ok I have resolved the conflicts.

**Pull request merged**
